diff -ru modules/ssl/mod_ssl.c modules/ssl/mod_ssl.c
--- modules/ssl/mod_ssl.c	2015-04-22 10:02:01.000000000 +0200
+++ modules/ssl/mod_ssl.c	2015-04-22 10:02:30.000000000 +0200
@@ -276,7 +276,7 @@
 #if defined(HAVE_TLS_ALPN) || defined(HAVE_TLS_NPN)
     SSL_CMD_SRV(AlpnPreference, ITERATE,
                 "Preference in Application-Layer Protocol Negotiation (ALPN), "
-                "protocols are chosed in the specified order")
+                "protocols are chosen in the specified order")
 #endif
     
     /* Deprecated directives. */
diff -ru modules/ssl/ssl_private.h modules/ssl/ssl_private.h
--- modules/ssl/ssl_private.h	2015-04-22 10:02:01.000000000 +0200
+++ modules/ssl/ssl_private.h	2015-04-22 10:03:31.000000000 +0200
@@ -177,12 +177,12 @@
 #endif
 
 /* ALPN Protocol Negotiation */
-#if OPENSSL_VERSION_NUMBER >= 0x10002000L && !defined(OPENSSL_NO_TLSEXT)
+#if defined(TLSEXT_TYPE_application_layer_protocol_negotiation) 
 #define HAVE_TLS_ALPN
 #endif
 
 /* Next Protocol Negotiation */
-#if !defined(OPENSSL_NO_NEXTPROTONEG) && !defined(OPENSSL_NO_TLSEXT) && defined(OPENSSL_NPN_NEGOTIATED)
+#if !defined(OPENSSL_NO_NEXTPROTONEG) && defined(OPENSSL_NPN_NEGOTIATED)
 #define HAVE_TLS_NPN
 #endif
 
